stl/inc/memory

```diff
@@ -173,22 +290,26 @@ namespace ranges {
         }
 
     private:
-        template <class _It1, class _Se1, class _It2, class _Se2>
-        _NODISCARD static uninitialized_move_result<_It1, _It2> _Uninitialized_move_unchecked(
-            _It1 _IFirst, const _Se1 _ILast, _It2 _OFirst, const _Se2 _OLast) {
-            _STL_INTERNAL_STATIC_ASSERT(input_iterator<_It1>);
-            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se1, _It1>);
-            _STL_INTERNAL_STATIC_ASSERT(_No_throw_forward_iterator<_It2>);
-            _STL_INTERNAL_STATIC_ASSERT(_No_throw_sentinel_for<_Se2, _It2>);
-            _STL_INTERNAL_STATIC_ASSERT(constructible_from<iter_value_t<_It2>, iter_rvalue_reference_t<_It1>>);
-
-            _Uninitialized_backout _Backout{_STD move(_OFirst)};
-
-            for (; _IFirst != _ILast && _Backout._Last != _OLast; ++_IFirst) {
-                _Backout._Emplace_back(_RANGES iter_move(_IFirst));
-            }
+        template <class _It, class _Se, class _Out, class _OSe>
+        _NODISCARD static uninitialized_move_result<_It, _Out> _Uninitialized_move_unchecked(
+            _It _IFirst, const _Se _ILast, _Out _OFirst, const _OSe _OLast) {
+            _STL_INTERNAL_STATIC_ASSERT(input_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(_No_throw_forward_iterator<_Out>);
+            _STL_INTERNAL_STATIC_ASSERT(_No_throw_sentinel_for<_OSe, _Out>);
+            _STL_INTERNAL_STATIC_ASSERT(constructible_from<iter_value_t<_Out>, iter_rvalue_reference_t<_It>>);
+
+            if constexpr (is_same_v<_Se, _It> && _Ptr_copy_cat<_It, _Out>::_Really_trivial) {
+                return _Copy_memmove_common(_IFirst, _ILast, _OFirst, _OLast);

```

Comment:
Here, `_Uninitialized_move_unchecked` is calling `_Copy_memmove_common`, but this is a false pattern. `memmove` means "handles overlapping ranges", not "has move semantics". I believe that overlapping is impossible for `uninitialized_meow`, so this should use `_Copy_memcpy_common`.

Reply:
This is what was there before. I want to note that `std::copy` also falls back to `_Copy_memmove` which confuses me to no end.

That said It will definitely break code if we change it to use (the faster) memcpy

Reply:
Removed the memmove helper and used memcpy version

Reply:
Summarizing our discussion on Discord - the difference between `std::copy` and `std::uninitialized_copy` is that `std::copy` handles overlapping in one direction, while `std::uninitialized_copy` forbids any overlapping - and similarly for `std::move` and `std::uninitialized_move`.

This _is_ confusing, so perhaps we should globally avoid `memcpy` and use `memmove` everywhere (@BillyONeal quoted the relevant assembly where we simply implement them identically so there is no performance difference).